src/lem/circuit.rs

```diff
@@ -708,13 +734,85 @@ impl Func {
+                        let src = bound_allocations.get(src)?;
+                        let tag = g.global_allocator.get_or_alloc_const(cs, tag.to_field())?;
+                        let allocated_ptr = AllocatedPtr::from_parts(tag, src.hash().clone());
+                        bound_allocations.insert(tgt.clone(), allocated_ptr);
+                    }
+                    Op::Add(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        // TODO check that the tags are correct
+                        let a_num = a.hash();
+                        let b_num = b.hash();
+                        let c_num = add(&mut cs.namespace(|| "add"), a_num, b_num)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, c_num);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Sub(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        // TODO check that the tags are correct
+                        let a_num = a.hash();
+                        let b_num = b.hash();
+                        let c_num = sub(&mut cs.namespace(|| "sub"), a_num, b_num)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, c_num);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Mul(tgt, a, b) => {
+                        let a = bound_allocations.get(a)?;
+                        let b = bound_allocations.get(b)?;
+                        // TODO check that the tags are correct
+                        let a_num = a.hash();
+                        let b_num = b.hash();
+                        let c_num = mul(&mut cs.namespace(|| "mul"), a_num, b_num)?;
+                        let tag = g
+                            .global_allocator
+                            .get_or_alloc_const(cs, Tag::Expr(Num).to_field())?;
+                        let c = AllocatedPtr::from_parts(tag, c_num);
+                        bound_allocations.insert(tgt.clone(), c);
+                    }
+                    Op::Div(_tgt, _a, _b) => {
+                        // TODO
+                    }
+                    Op::Hide(tgt, _sec, _pay) => {
+                        // TODO
                         let allocated_ptr = AllocatedPtr::from_parts(
-                            g.global_allocator.get_or_alloc_const(cs, tag.to_field())?,
+                            g.global_allocator.get_or_alloc_const(cs, F::ZERO)?,
```

Comment:
Maybe `todo!()` rather than `F::ZERO` — unless you need to actually successfully run this with bad values. I don't want us to condition ourselves (as either authors or reviewers) to get used to seeing these wrong values in the code. Especially since there actually are some nearby places (like in `commit`) that a literal zero *could* belong (though I think we've moved to making this a more specific constant for exactly that reason).

Reply:
`todo!()` will make the eval test fail. I will just leave it empty until we can fix this